Release 1

SaifKhalifa opened the pull request:
This is the final solution with all the requirements, i have only used the strategy pattern since it's the only one i have found suitable.

SaifKhalifa committed:
Add .NET 8.0 project and solution files

Created `weather-bot-monitoring.csproj` with .NET SDK structure targeting .NET 8.0, including essential properties for output type, root namespace, and nullable reference types. Added `weather-bot-monitoring.sln` to define the solution structure for Visual Studio with configurations for Debug and Release modes.

SaifKhalifa committed:
Add WeatherData model class with location and weather properties

Introduced a new class `WeatherData` that encapsulates
weather-related information, including properties for
`Location`, `Temperature`, and `Humidity`.

SaifKhalifa committed:
Add IWeatherBot interface for weather data activation

SaifKhalifa committed:
Add RainBot, SnowBot, and SunBot classes implementing IWeatherBot for weather condition activation

SaifKhalifa committed:
Refactor WeatherData class to include namespace and improve structure

SaifKhalifa committed:
Update CheckAndActivate method to be of type boolean in IWeatherBot and implement return value in RainBot, SnowBot, and SunBot

SaifKhalifa committed:
Removed brackets from namespaces to reduce indentation

SaifKhalifa committed:
Add IWeatherDataParser interface and implementations for JSON and XML parsing

SaifKhalifa committed:
Add sample weather data in JSON and XML formats

SaifKhalifa committed:
Add BotsConfig class with configurations for RainBot, SunBot, and SnowBot

SaifKhalifa committed:
Refactor RainBot, SnowBot, and SunBot to accept message parameter and change threshold type to int

SaifKhalifa committed:
Add WeatherBotsConfig.json with configuration settings for RainBot, SunBot, and SnowBot

SaifKhalifa committed:
Add main program for weather bot monitoring with data input and bot activation

SaifKhalifa committed:
Remove debug output for bot configuration loading in Program.cs

OmarAbbadi01 commented on Parsers/XmlWeatherDataParser.cs:
```diff
@@ -0,0 +1,12 @@
+    {
+        var serializer = new XmlSerializer(typeof(WeatherData));
+        using var reader = new StringReader(input);
+        return (WeatherData)serializer.Deserialize(reader);
```
Isn't there an async version of this method?

SaifKhalifa replied:
Resolved

OmarAbbadi01 commented on Parsers/JsonWeatherDataParser.cs:
```diff
@@ -0,0 +1,11 @@
+{
+    public WeatherData Parse(string input)
+    {
+        return JsonSerializer.Deserialize<WeatherData>(input);
```
Isn't there an async version of this method?

SaifKhalifa replied:
Resolved

OmarAbbadi01 commented on Program.cs:
```diff
@@ -0,0 +1,113 @@
+        }
+
+        // Choose parser
+        IWeatherDataParser parser = format == "1"
```
Move this to a factory class following the Factory Design Pattern

SaifKhalifa replied:
Resolved

OmarAbbadi01 commented on Program.cs:
```diff
@@ -0,0 +1,113 @@
+        }
+        else
+        {
+            Console.WriteLine("At least one bot was activated.");
```
Can you write here which bot(s) was/were activated?
HINT: Use LINQ to filter `bots`

SaifKhalifa replied:
Resolved

OmarAbbadi01 commented on Program.cs:
```diff
@@ -0,0 +1,113 @@
+        try
+        {
+            string configContent = File.ReadAllText("WeatherBotsConfig.json");
+            config = JsonSerializer.Deserialize<BotsConfig>(configContent, new JsonSerializerOptions
```
Try this async flow

```
using FileStream stream = File.OpenRead(filePath);
var x = await JsonSerializer.DeserializeAsync<X>(stream);
```

SaifKhalifa replied:
Resolved with factory and async parsers update!

SaifKhalifa committed:
Add IWeatherDataParserFactory interface and implementation

using factory pattern

SaifKhalifa committed:
Refactor IWeatherDataParser for async support

SaifKhalifa committed:
Refactor Program.cs for async parsing and error handling

- Made the Main method asynchronous to support await.
- Refactored parser selection to use WeatherDataParserFactory.
- Updated weather data parsing to use ParseAsync for non-blocking I/O.
- Improved error handling with more descriptive messages.

SaifKhalifa committed:
Use async file reading

SaifKhalifa committed:
Refactor bot activation logic to print activated bots

SaifKhalifa committed:
Main prgram refactor: moved choosing input format to seperate method

SaifKhalifa committed:
Main program refactor: moved weather data input to a seperate method

SaifKhalifa committed:
Main program refactor: moved getting parser to a seperate method.

SaifKhalifa committed:
Main program refactor: Moved weather data parsing logic to a seperate method.

SaifKhalifa committed:
Main program refactor: moved bot config loading to a seperte method.

SaifKhalifa committed:
Main prgoram refactor: moved creating bots logic to a seperate method

SaifKhalifa committed:
Main program refactor: moved running bots logic to a seperate method.

Main program refactor completed and it's now easier to maintain and easier for testing and mocking.

SaifKhalifa commented:
All changes should be resolved by now, and main program file is refactored to be easier for maintain and easier for testing and mocking
